Restructure pipelines for verbosity

AyanSinhaMahapatra opened the pull request:
Remove scan_codebase_packages pipeline, and restructure inspect_packages pipeline into load_sbom and resolve_packages pipelines.

Reference: https://github.com/nexB/scancode.io/issues/1040
Reference: https://github.com/nexB/scancode.io/discussions/1035
Reference: https://github.com/nexB/scancode.io/issues/1034
Reference: #1044

tdruez commented on scanpipe/pipelines/load_sbom.py:
```diff
@@ -0,0 +1,78 @@
+        )
+
+    def get_sbom_inputs(self):
+        """Locate all the SBOMs from the project's input/ directory."""
```
This description needs an update. It loads the detected manifest file but not from the project's input/ directory

AyanSinhaMahapatra replied:
Updated and improved all related docstrings, thanks!

tdruez commented on scanpipe/pipelines/load_sbom.py:
```diff
@@ -0,0 +1,78 @@
+            )
+            return
+
+        for resource in self.manifest_resources:
+            if packages := resolve.resolve_packages(
+                input_location=resource.location,
+                package_registry=resolve.sbom_registry,
+            ):
+                for package_data in packages:
+                    package_data = resolve.set_license_expression(package_data)
+                    update_or_create_package(self.project, package_data)
+            else:
+                self.project.add_error(
+                    description="No packages could be resolved for SBOM",
+                    model="get_packages_from_sboms",
+                    details={"path": resource.path},
+                )
```
This logic belongs to a pipe, we want to avoid the code complexity on the pipeline classes.
Also, it needs refactoring, to help with testing and readability.

AyanSinhaMahapatra replied:
Refactored and moved this to a pipe, thanks!

tdruez commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,88 @@
+
+class ResolveDependencies(ScanCodebase):
+    """
+    Inspect a codebase manifest files and resolve their associated packages.
```
We need to clarify "a codebase manifest files"

AyanSinhaMahapatra replied:
Updated and improved this and other pipelines, thanks!

tdruez commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,88 @@
+        )
+
+    def get_manifest_inputs(self):
+        """Locate all the manifest files from the project's input/ directory."""
```
Same comment as above.

AyanSinhaMahapatra replied:
Updated and improved this and all related docstrings, thanks!

tdruez commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,88 @@
+        """Locate all the manifest files from the project's input/ directory."""
+        self.manifest_resources = resolve.get_manifest_resources(self.project)
+
+    def get_packages_from_manifest(self):
+        """Get packages data from manifest files."""
+        self.resolved_packages = []
+
+        if not self.manifest_resources.exists():
+            self.project.add_warning(
+                description="No manifests found for resolving packages",
+                model="get_packages_from_manifest",
+            )
+            return
+
+        for resource in self.manifest_resources:
+            if packages := resolve.resolve_packages(
+                input_location=resource.location,
+                package_registry=resolve.resolver_registry,
+            ):
+                self.resolved_packages.extend(packages)
+            else:
+                self.project.add_error(
+                    description="No packages could be resolved for",
+                    model="get_packages_from_manifest",
+                    details={"path": resource.path},
```
Same comment as above.
Also, this seems duplicated with the get_packages_from_sboms, something to take in account while refactoring.

tdruez commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,88 @@
+                    details={"path": resource.path},
+                )
+
+    def create_resolved_packages(self):
+        """Create the resolved packages and their dependencies in the database."""
+        for package_data in self.resolved_packages:
+            package_data = resolve.set_license_expression(package_data)
+            dependencies = package_data.pop("dependencies", [])
+            update_or_create_package(self.project, package_data)
+
+            for dependency_data in dependencies:
+                resolved_package = dependency_data.get("resolved_package")
+                if resolved_package:
+                    resolved_package.pop("dependencies", [])
+                    update_or_create_package(self.project, resolved_package)
```
Move the logic on pipes.

tdruez reviewed (changes requested):
@AyanSinhaMahapatra see my various comments.

Also, the following is missing:
- Unit tests
- Release notes (very important considering the major pipeline renaming, changes, and additions)
- The Built-in pipe docs need to be updated with those changes

AyanSinhaMahapatra commented:
@tdruez thanks for the review comments!
I've updated the PR to address all the issues mentioned above.
Didn't update the docs/changelog before as I wanted you and @pombredanne to check out the pipeline restructuring and approve, but I've updated them now, and this is ready for another round of review. Thanks!

AyanSinhaMahapatra commented:
See also comment at https://github.com/nexB/scancode.io/discussions/1035#discussioncomment-8423854, we've addressed all the comments there and are adhering to the naming convention now for all pipelines.

tdruez commented:
@AyanSinhaMahapatra Thanks for the changes. Everything looks fine on my side but I'd like to get @pombredanne review as well before merging as this is quite a big change:
- 2 new pipelines
- 1 removed pipeline

pombredanne commented on scanpipe/pipelines/load_sbom.py:
```diff
@@ -21,15 +21,17 @@
+class LoadSBOM(ScanCodebase):
     """
-    Scan a codebase for PURLs without assembling full packages/dependencies.
+    Detects and loads all associated packages from SBOMs.
```
We need to keep the same imperative style across: 

```suggestion
    Load package data from one or more SBOMs.
```

pombredanne commented on scanpipe/pipelines/load_sbom.py:
```diff
@@ -21,15 +21,17 @@
+    Supported BOMs:
+    - SPDX document
+    - CycloneDX BOM
+    - AboutCode ABOUT file
```
```suggestion
    Other formats:
    - AboutCode .ABOUT files for package curations.  
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+
+class ResolveDependencies(ScanCodebase):
+    """
+    Resolve dependencies from package manifests/lockfiles.
```
```suggestion
    Resolve dependencies from packages manifests and lockfiles.
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+    Resolve dependencies from package manifests/lockfiles.
+
+    This pipeline inspects codebase lockfiles/manifest files
+    which contain package requirements and resolve their
```
```suggestion
    that contain dependencies requirements and resolve these
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+    """
+    Resolve dependencies from package manifests/lockfiles.
+
+    This pipeline inspects codebase lockfiles/manifest files
```
```suggestion
    This pipeline collects lockfiles and manifest files
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+
+    This pipeline inspects codebase lockfiles/manifest files
+    which contain package requirements and resolve their
+    associated packages.
```
```suggestion
    to a concrete set of package versions.
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+    which contain package requirements and resolve their
+    associated packages.
+
+    Supports resolved packages for:
```
```suggestion
    Supports resolving packages for:
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+    associated packages.
+
+    Supports resolved packages for:
+    - Python: using nexB/python-inspector, supports requirements.txt and
```
```suggestion
    - Python: using python-inspector, using requirements.txt and
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+
+    Supports resolved packages for:
+    - Python: using nexB/python-inspector, supports requirements.txt and
+    setup.py manifests as input
```
```suggestion
    setup.py manifests as inputs.
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+
+    def get_manifest_inputs(self):
+        """
+        Locate all the package manifest files for which package resolvers are
```
```suggestion
        Locate package manifest files with a supported package resolver.
```

pombredanne commented on scanpipe/pipelines/resolve_dependencies.py:
```diff
@@ -0,0 +1,77 @@
+    def get_manifest_inputs(self):
+        """
+        Locate all the package manifest files for which package resolvers are
+        supported, in the codebase resources.
```
```suggestion
```

pombredanne commented on scanpipe/pipelines/scan_codebase.py:
```diff
@@ -28,15 +28,11 @@
-    directory.
+    This pipeline does not further scan the files contained in a package
+    for license and copyrights and only considers the declared license
+    of a package.
```
```suggestion
    of a package. It does not scan for system (Linux distro) packages.
```

pombredanne commented on scanpipe/pipelines/scan_single_package.py:
```diff
@@ -36,7 +36,10 @@
 class ScanSinglePackage(Pipeline):
     """
-    Scan a single package file or package archive with ScanCode-toolkit.
+    Scan a single package file or package archive.
```
```suggestion
    Scan a single package archive (or package manifest file).
```

pombredanne reviewed (changes requested):
LGTM!
I posted a few nits regarding wordsmithing.
Please review, apply the good ones and this is good to merge.

AyanSinhaMahapatra commented:
@pombredanne thanks for the suggestions, I've added them all with some more docstring updates!

pombredanne commented on docs/built-in-pipelines.rst:
```diff
@@ -126,14 +142,6 @@ Scan Codebase
 
-.. _pipeline_scan_codebase_package:
-
-Scan Codebase Package
```
Since this is renamed... we may still need an entry in the doc? This can be added later.

AyanSinhaMahapatra replied:
@pombredanne we do have the new pipelines in the docs correctly, see: https://github.com/nexB/scancode.io/blob/restructure-pipelines/docs/built-in-pipelines.rst?plain=1#L61

Is this what you meant?

We are not having entries on renamed pipelines in the docs anymore: see https://github.com/nexB/scancode.io/pull/1053/files#diff-8a1644efb2ff73c9d7a85c18c73d2558528056d2fe7bb2adafe2502df72f4e2e also. The old pipelines are available only through the API/CLI for compatibility where they are renamed to the new pipelines. Otherwise we are not having the old pipeline names anywhere.

pombredanne reviewed (approved):
Thanks.... just one last nit, but nothing in the way of merging.

tdruez commented:
@AyanSinhaMahapatra Merged, thanks!
